mopro-core/src/middleware/gpu_explorations/metal/msm.rs

```diff
@@ -0,0 +1,359 @@
+use ark_bn254::{Fq, Fr as ScalarField, G1Affine as GAffine, G1Projective as G};
+use ark_ec::{AffineRepr, Group, VariableBaseMSM};
+use ark_ff::PrimeField;
+use ark_std::{cfg_into_iter, vec::Vec, One, Zero};
+
+// For benchmarking
+use std::time::{Duration, Instant};
+
+use crate::middleware::gpu_explorations::metal::abstraction::{
+    errors::MetalError,
+    limbs_conversion::{FromLimbs, ToLimbs},
+    state::*,
+};
+use crate::middleware::gpu_explorations::utils::{benchmark::BenchmarkResult, preprocess};
+
+use metal::*;
+use objc::rc::autoreleasepool;
+
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+// Helper function for getting the windows size
+fn ln_without_floats(a: usize) -> usize {
+    // log2(a) * ln(2)
+    (ark_std::log2(a) * 69 / 100) as usize
+}
+
+pub fn metal_msm(points: &[GAffine], scalars: &[ScalarField]) -> Result<G, MetalError> {
+    let modulus_bit_size = ScalarField::MODULUS_BIT_SIZE as usize;
+    let zero = G::zero();
+    let one = ScalarField::one();
+
+    let instances_size = ark_std::cmp::min(points.len(), scalars.len());
+    let window_size = if instances_size < 32 {
+        3
+    } else {
+        ln_without_floats(instances_size) + 2
+    };
+    let buckets_size = (1 << window_size) - 1;
+    let window_starts: Vec<usize> = (0..modulus_bit_size).step_by(window_size).collect();
+
+    // flatten scalar and base to Vec<u32> for GPU usage
+    let scalars_limbs = cfg_into_iter!(scalars)
+        .map(|s| s.into_bigint().to_u32_limbs())
+        .flatten()
+        .collect::<Vec<u32>>();
+    let bases_limbs = cfg_into_iter!(points)
+        .map(|b| {
+            let b = b.into_group();
+            b.x.to_u32_limbs()
+                .into_iter()
+                .chain(b.y.to_u32_limbs())
+                .chain(b.z.to_u32_limbs())
+                .collect::<Vec<_>>()
+        })
+        .flatten()
+        .collect::<Vec<u32>>();
+    let buckets_matrix_limbs = {
+        // buckets_size * num_windows is for parallelism on windows (variable-based MSM)
+        let matrix = vec![zero; buckets_size * window_starts.len()];
+        cfg_into_iter!(matrix)
+            .map(|b| {
+                b.x.to_u32_limbs()
+                    .into_iter()
+                    .chain(b.y.to_u32_limbs())
+                    .chain(b.z.to_u32_limbs())
+                    .collect::<Vec<_>>()
+            })
+            .flatten()
+            .collect::<Vec<u32>>()
+    };
```

Comment:
Possible optimization direction: enhance speed for preparing inputs for GPU

Reply:
can you open a new issue?
I think it can be outside of this PR
(and because this PR is too big)